### PR TITLE
IWF-163: Add retry policy to QueryWorkflow calls

### LINKED ISSUE #437: Refactor persistence_test and move API call retry to API Service
Under certain circumstances (starting a continuedAsNew workflow is one of them) it might take time before Temporal/Candace API calls can be resolved successfully and expected data is returned. We should refactor the code and retry API calls when a certain response is received (for example: 500 response)

**ACs:**

- Retry API calls when certain responses are received
  - TODO: Need to decide on checking error types, and backoff strategy
- Refactor [persistence_test.go](https://github.com/indeedeng/iwf/blob/main/integ/persistence_test.go) so it does not need to use Sleep.time method

### PR DESCRIPTION
**Pull request by lwolczynski**
### Checklist
- [ ] Code compiles correctly
- [ ] Tests for the changes have been added
- [ ] All tests passing
- [ ] **This PR change is backwards-compatible**
- [ ] **This PR CONTAINS a (planned) breaking change** (it is not backwards compatible)

### Related Issue
Closes #437

### CODE REVIEW

File: cmd/server/iwf/iwf.go [hunk 1]
```diff
@@ -124,7 +124,11 @@ func start(c *cli.Context) {
 		if err != nil {
 			rawLog.Fatalf("Unable to connect to Temporal because of error %v", err)
 		}
-		unifiedClient = temporalapi.NewTemporalClient(temporalClient, config.Interpreter.Temporal.Namespace, converter.GetDefaultDataConverter(), false)
+		retryPolicy := temporalapi.QueryWorkflowFailedRetryPolicy{
+			InitialIntervalSeconds: config.GetInitialIntervalSecondsWithDefault(),
+			MaximumAttempts:        config.GetMaximumAttemptsWithDefault(),
+		}
+		unifiedClient = temporalapi.NewTemporalClient(temporalClient, config.Interpreter.Temporal.Namespace, converter.GetDefaultDataConverter(), false, retryPolicy)
```
Comment (longquanzheng) [inline cmd/server/iwf/iwf.go:131]: If you have `config.GetInitialIntervalSecondsWithDefault()`, then the expectation is that the code that need to use the value will call it to get the value with default, instead of passing into here. 

So it will be `temporalapi.NewTemporalClient(temporalClient, config.Interpreter.Temporal.Namespace, converter.GetDefaultDataConverter(), false, config.ApiConfig.QueryRetryPolicy)`

Also consider using pointers to make this optional and load the default values in `client.go`.
Basically the config can be defined it as a pointer instead of struct. So that internally, we just need to pass a `nil` into this, which should let it use the default value.

File: service/client/cadence/client.go [hunk 2]
```diff
@@ -227,9 +240,24 @@ func (t *cadenceClient) ListWorkflow(
...
-		return err
+	var qres encoded.Value
+	var err error
+
+	attempt := 1
+	for attempt <= t.queryWorkflowFailedRetryPolicy.MaximumAttempts {
+		qres, err = queryWorkflowWithStrongConsistency(t, ctx, workflowID, runID, queryType, args)
+		if err != nil {
+			if t.isQueryFailedError(err) {
+				if attempt == t.queryWorkflowFailedRetryPolicy.MaximumAttempts {
```
Comment (longquanzheng) [inline service/client/cadence/client.go:251]: Nit: I don't think we need another check here, since it'a already checked by the for loop condition?

Although we will have to check `err != nil` again outside of the loop

Comment (lwolczynski) [reply service/client/cadence/client.go:251]: Boils down to the same thing, but it might be a little cleaner. Done

File: service/client/temporal/client.go [hunk 3]
```diff
@@ -20,23 +21,45 @@ import (
...
+	queryWorkflowFailedRetryPolicy QueryWorkflowFailedRetryPolicy
 }
 
 func NewTemporalClient(
-	tClient client.Client, namespace string, dataConverter converter.DataConverter, memoEncryption bool,
+	tClient client.Client, namespace string, dataConverter converter.DataConverter, memoEncryption bool, retryPolicy *config.QueryWorkflowFailedRetryPolicy,
 ) uclient.UnifiedClient {
+	var rp QueryWorkflowFailedRetryPolicy
+
+	if retryPolicy.InitialIntervalSeconds == 0 {
```
Comment (lwolczynski) [inline service/client/temporal/client.go:45]: If I'm not mistaken, not providing `Config.ApiConfig.QueryWorkflowFailedRetryPolicy` in yaml file will make the default int value show (which is 0). In other words, we shouldn't need a nil check here
